## Lesson 01
### Linear Regression. Gradient Descent.

In [1]:
import numpy as np
import matplotlib.pyplot as plt

In [2]:
# Experience, years
X0 = np.array([1, 1, 2, 1, 3, 0, 5, 10, 1, 2])
X0

array([ 1,  1,  2,  1,  3,  0,  5, 10,  1,  2])

In [3]:
# Intercept (array of '1' values)
ones = np.ones_like(X0)
# Concatenate
X = np.append([ones], [X0], axis=0)
X

array([[ 1,  1,  1,  1,  1,  1,  1,  1,  1,  1],
       [ 1,  1,  2,  1,  3,  0,  5, 10,  1,  2]])

In [4]:
# Scores (target)
y = [45, 55, 50, 59, 65, 35, 75, 80, 50, 60]

In [5]:
# MAE (Mean Absolute Error)
def calc_mae(y, y_pred):
  err = np.mean(np.abs(y - y_pred))
  return err

In [6]:
# MSE (Mean Squared Error)
def calc_mse(y, y_pred):
  err = np.mean((y - y_pred)**2)
  return err

#### Alpha and number of iterations for Gradient Descend

In [7]:
n = X.shape[1]
W = np.array([1, 0.5])
alpha = 1e-2
iterations = 1201
n, W, alpha, iterations

(10, array([1. , 0.5]), 0.01, 1201)

In [8]:
for i in range(iterations):
  y_pred = np.dot(W, X)
  err = calc_mse(y, y_pred)
  for ii in range(W.shape[0]):
    W[ii] -= alpha * (1/n * 2 * np.sum(X[ii] * (y_pred - y)))
  if i % 200 == 0:
    print(i, W, err)

0 [2.102 3.9  ] 3173.15
200 [41.83683774  4.90699865] 61.9177717428135
400 [46.56511152  4.03388672] 46.18175564810758
600 [47.14967657  3.92594232] 45.941233404700036
800 [47.2219474   3.91259695] 45.937557064435396
1000 [47.23088237  3.91094704] 45.937500872219886
1200 [47.23198702  3.91074306] 45.93750001333172


MSE stopped changing much at the value of `~45.9375`